### PR TITLE
SNS-Low Power Current

### PR DESCRIPTION
**Pull request by sa-fx**
Feedback greatly appreciated

### CODE REVIEW

File: lib/sensors/low_power_current.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include <low_power_current.hpp>
+
+namespace hyped::sensors {
+
+std::optional<LowPowerCurrent> LowPowerCurrent::create(core::ILogger &logger,
+                                                       io::II2c &i2c,
+                                                       std::uint8_t channel = kdefault_i2c_address)
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:7]: `const`

Comment (maxguy2001) [reply lib/sensors/low_power_current.cpp:7]: also `kDefaultI2cAddress`

Comment (maxguy2001) [reply lib/sensors/low_power_current.cpp:7]: The [google style guide](https://google.github.io/styleguide/cppguide.html) might help with standard practice/naming style.

File: lib/sensors/low_power_current.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+  return LowPowerCurrent(logger, i2c, channel);
+}
+
+LowPowerCurrent::LowPowerCurrent(core::ILogger &logger, io::II2c &i2c, std::uint8_t channel)
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:12]: `const`?

File: lib/sensors/low_power_current.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    : logger_(logger),
+      channel_(channel),
+      i2c_(i2c){};
+
+LowPowerCurrent::~LowPowerCurrent(){};
+
+std::optional<float> LowPowerCurrent::readCurrent()
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:19]: `core::Float`

File: lib/sensors/low_power_current.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+{
+  if (!(i2c_.readByte(channel_, kcurrent_reg))) {
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:21]: style guide...

File: lib/sensors/low_power_current.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    logger_.log(core::LogLevel::kFatal, "Failed to read current on channel %d", channel_);
+    return std::nullopt;
+  }
+  logger_.log(core::LogLevel::kDebug, "Current read on channel %d successful", channel_);
+  return i2c_.readByte(channel_, kcurrent_reg);
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:26]: style guide ...

File: lib/sensors/low_power_current.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#include <cstdint>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+// The datasheet lists four possible I2C addresses; 0x40 has been defined as the default
+static constexpr std::uint8_t kdefault_i2c_address = 0x40;
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.hpp:12]: style guide violations here ...

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/low_power_current.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+std::optional<core::Float> LowPowerCurrent::readCurrent()
+{
+  const auto byte = i2c_->readByte(channel_, kCurrentReg);
+  if (!byte) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read current on channel %d", channel_);
+    return std::nullopt;
+  }
+  logger_.log(core::LogLevel::kDebug, "Current read on channel %d successful", channel_);
+  // Return current in Amps
+  return (core::Float)byte.value() / kAmpsDivisor;
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:31]: `static_cast<core::Float>(...);`

Comment (maxguy2001) [reply lib/sensors/low_power_current.cpp:31]: Also is there a better name than `kAMpsDivisor`?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]
Looking much better!

File: lib/sensors/low_power_current.cpp (hunk 7 above)
Comment (sa-fx) [reply lib/sensors/low_power_current.cpp:31]: @SnickeyX

File: lib/sensors/low_power_current.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+std::optional<core::Float> LowPowerCurrent::readCurrent()
+{
+  const auto byte = i2c_->readByte(channel_, kCurrentReg);
+  if (!byte) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read current on channel %d", channel_);
+    return std::nullopt;
+  }
+  logger_.log(core::LogLevel::kDebug, "Current read on channel %d successful", channel_);
+  // Divide milliAmps output by 1,000 to return Amps value
+  return static_cast<core::Float>(byte.value() / 1000);
```
Comment (maxguy2001) [inline lib/sensors/low_power_current.cpp:31]: If `byte.value()` returns an integer type then you're just doing integer division and then casting so should it not be something along the lines of:
```
return static_cast<core::Float>(byte.value())/1000;
```
?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/low_power_current.cpp (hunk 8 above)
Comment (TomLonergan03) [inline lib/sensors/low_power_current.cpp:31]: dereference an optional using `*` not `.value()` ie `*byte`

File: lib/sensors/low_power_current.hpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace hyped::sensors {
+
+// The datasheet lists four possible I2C addresses; 0x40 has been defined as the default
+constexpr std::uint8_t kDefaultI2cAddress = 0x40;
+constexpr std::uint8_t kI2cAddress2       = 0x41;
+constexpr std::uint8_t kI2cAddress3       = 0x44;
+constexpr std::uint8_t kI2cAddress4       = 0x45;
+// Registers with reference to the INA219 driver code
+constexpr std::uint8_t kCurrentReg = 0x04;
```
Comment (TomLonergan03) [inline lib/sensors/low_power_current.hpp:18]: `kCurrentRegister`

File: lib/sensors/low_power_current.cpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace hyped::sensors {
+
+std::optional<LowPowerCurrent> LowPowerCurrent::create(core::ILogger &logger,
+                                                       std::shared_ptr<io::II2c> i2c,
+                                                       const std::uint8_t channel
+                                                       = kDefaultI2cAddress)
+{
+  return LowPowerCurrent(logger, i2c, channel);
+}
```
Comment (TomLonergan03) [inline lib/sensors/low_power_current.cpp:11]: if this has no failure state then there is no need for the create function

Comment (TomLonergan03) [reply lib/sensors/low_power_current.cpp:11]: if there is configuration required that is yet to be implemented then ignore me

File: lib/sensors/low_power_current.cpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+LowPowerCurrent::LowPowerCurrent(core::ILogger &logger,
+                                 std::shared_ptr<io::II2c> i2c,
+                                 const std::uint8_t channel)
+    : logger_(logger),
+      channel_(channel),
+      i2c_(i2c){};
```
Comment (TomLonergan03) [inline lib/sensors/low_power_current.cpp:18]: remove the `;` after all the methods in the file

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/sensors/low_power_current.cpp (hunk 10 above)
Comment (SnickeyX) [reply lib/sensors/low_power_current.cpp:12]: ^^ TODOLater

File: lib/sensors/low_power_current.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#include <low_power_current.hpp>
+
+namespace hyped::sensors {
+
+// TODO: Implement failure state for create
+std::optional<LowPowerCurrent> LowPowerCurrent::create(core::ILogger &logger,
+                                                       std::shared_ptr<io::II2c> i2c,
+                                                       const std::uint8_t channel
+                                                       = kDefaultI2cAddress)
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:9]: dont do default here, check it in `create` itself, also `defaultLowPowerCurrentAddress`

File: lib/sensors/low_power_current.cpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+{
+  return LowPowerCurrent(logger, i2c, channel);
+}
+
+LowPowerCurrent::LowPowerCurrent(core::ILogger &logger,
+                                 std::shared_ptr<io::II2c> i2c,
+                                 const std::uint8_t channel)
+    : logger_(logger),
+      channel_(channel),
+      i2c_(i2c)
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:19]: switch these around so they are in the order of arguements

File: lib/sensors/low_power_current.hpp [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+#include <core/types.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+// The datasheet lists four possible I2C addresses; 0x40 has been defined as the default
+constexpr std::uint8_t kDefaultI2cAddress = 0x40;
+constexpr std::uint8_t kI2cAddress2       = 0x41;
+constexpr std::uint8_t kI2cAddress3       = 0x44;
+constexpr std::uint8_t kI2cAddress4       = 0x45;
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.hpp:17]: `LowPowerAddressX` where x in [2,4]

File: lib/sensors/low_power_current.hpp [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+ private:
+  LowPowerCurrent(core::ILogger &logger, std::shared_ptr<io::II2c> i2c, const std::uint8_t channel);
+
+ private:
+  core::ILogger &logger_;
+  std::shared_ptr<io::II2c> i2c_;
+  const std::uint8_t channel_;
+};
+}  // namespace hyped::sensors
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.hpp:40]: space before last bracket

**Review by SnickeyX — CHANGES_REQUESTED** [review]

File: lib/sensors/low_power_current.cpp (hunk 12 above)
Comment (SnickeyX) [reply lib/sensors/low_power_current.cpp:9]: also maybe call `channel` - `device_address`

Comment (sa-fx) [reply lib/sensors/low_power_current.cpp:9]: Isn't `kDefaultLowPowerCurrentAddress` a bit long? Mind you, not sure what would be a suitably shorter yet equally clear substitute

File: lib/sensors/low_power_current.hpp (hunk 14 above)
Comment (sa-fx) [reply lib/sensors/low_power_current.hpp:17]: These are defined purely for the sake of documenting what the addresses are from the datasheet. They aren't actually used anywhere else in the code. Therefore, should these be here? Or are they completely redundant and should be removed?

File: lib/sensors/low_power_current.cpp (hunk 12 above)
Comment (SnickeyX) [reply lib/sensors/low_power_current.cpp]: @kshxtij said do it to be more clear

**Review by maxguy2001 — APPROVED** [review]

File: lib/sensors/low_power_current.cpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  }
+  logger_.log(core::LogLevel::kDebug, "Current read on channel %d successful", device_address_);
+  // Divide milliAmps output by 1,000 to return Amps value
+  return static_cast<core::Float>(*byte) / 1000;
+}
+
+std::uint8_t LowPowerCurrent::getDeviceAddress() const
+{
+  return device_address_;
+}
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:41]: We dont need this function

**Review by SnickeyX — CHANGES_REQUESTED** [review]

File: lib/sensors/low_power_current.cpp [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+LowPowerCurrent::~LowPowerCurrent()
+{
+}
+
+std::optional<core::Float> LowPowerCurrent::readCurrent()
+{
+  const auto byte = i2c_->readByte(device_address_, kLowPowerCurrentRegister);
+  if (!byte) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read current on channel %d", device_address_);
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:30]: channel? same with the comment below. I think "Failed to read low power current" should be fine

File: lib/sensors/low_power_current.cpp [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    return std::nullopt;
+  }
+  logger_.log(core::LogLevel::kDebug, "Current read on channel %d successful", device_address_);
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:33]: "Successfully read low power current"

File: lib/sensors/low_power_current.cpp [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+{
+}
+
+LowPowerCurrent::~LowPowerCurrent()
+{
+}
+
+std::optional<core::Float> LowPowerCurrent::readCurrent()
+{
+  const auto byte = i2c_->readByte(device_address_, kLowPowerCurrentRegister);
```
Comment (SnickeyX) [inline lib/sensors/low_power_current.cpp:28]: `low_power_current` maybe, byte by itself is a bit ambiguous

**Review by SnickeyX — CHANGES_REQUESTED** [review]

**Review by TomLonergan03 — APPROVED** [review]

**Review by SnickeyX — APPROVED** [review]
:)

**Review by kshxtij — APPROVED** [review]

**Pull request merged**